chirp_client.cc

```diff
@@ -0,0 +1,302 @@
+  grpc::Status status = stub_->monitor_check(&context, request, &reply);
+  
+  // User doesn't exist
+  if(status.error_code() == grpc::StatusCode::NOT_FOUND) {
+    std::cout << "Failed monitor for user '" << user << "'" << std::endl;
+    std::cout << status.error_message() << std::endl;
+    return false;
+  }
+  // Warn user if they don't follow anyone
+  if(status.error_code() == grpc::StatusCode::UNIMPLEMENTED) {
+    std::cout << "Warning: user '" << user << "' " << status.error_message() << std::endl;
+    return true;
+  }
+  // Tell user who they follow
+  if(status.ok()) {
+    std::cout << "Currently following: " << reply.names() << std::endl;
+    return true;
+  }
+}
+
+void ChirpClient::PrintChirpThread(std::vector<chirp::Chirp> &reply_chirps, bool isThread) {
+  // only need to print one chirp
+  if(!isThread) {
+    std::cout << "|----------------------------------|" << std::endl;
+    std::cout << "  [" << reply_chirps.at(0).username() << "]" << std::endl;
+    std::cout << "  ID: " << reply_chirps.at(0).id(); 
+    if(!reply_chirps.at(0).parent_id().empty()) {
+        std::cout << ", (reply to: " << reply_chirps.at(0).parent_id() << ")";
+    }
+    std::cout << std::endl;
+    std::cout << "  - '" << reply_chirps.at(0).text()  << "'" << std::endl;
+    std::cout << "  " << google::protobuf::util::TimeUtil::ToString(google::protobuf::util::TimeUtil::SecondsToTimestamp(reply_chirps.at(0).timestamp().seconds())) << std::endl;
+    std::cout << "|----------------------------------|" << std::endl;
+  }
+
+  // Is a thread so we need to tab correctly
+  else {
+    std::string tab = "\t";
+    std::string total_tabs = "";
+    for(chirp::Chirp chirp : reply_chirps) {
+      total_tabs = "";
+
+      for(int i = 1; i < chirp.depth(); ++i) {
+        total_tabs += tab;
+      }
+      std::cout << total_tabs << "|----------------------------------|" << std::endl;
+      std::cout << total_tabs << "  [" << chirp.username() << "]" << std::endl;
+      std::cout << total_tabs <<"  ID: " << chirp.id();
+      if(!chirp.parent_id().empty()) {
+        std::cout << ", (reply to ID: " << chirp.parent_id() << ")";
+      }
+      std::cout << std::endl;
+      std::cout << total_tabs << "  - '" << chirp.text() << "'" << std::endl;
+      std::cout << total_tabs << "  " << google::protobuf::util::TimeUtil::ToString(google::protobuf::util::TimeUtil::SecondsToTimestamp(chirp.timestamp().seconds())) << std::endl;
+      std::cout << total_tabs << "|----------------------------------|" << std::endl;
+    }
+  }
+
+}
+// Instantiate client. It requires a channel, out of which the actual RPCs
+// are created. This channel models a connection to an endpoint (in this case,
+// localhost at port 50000). We indicate that the channel isn't authenticated
+// (use of InsecureChannelCredentials()).
+int main(int argc, char** argv) {
+  ChirpClient greeter(grpc::CreateChannel("localhost:50000", grpc::InsecureChannelCredentials()));
```

Comment:
Greeter probably isn't a meaningful name for this application.

Reply:
true